palette/moreland/convert.go

```diff
@@ -0,0 +1,228 @@
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package moreland
+
+import (
+	"fmt"
+	"image/color"
+	"math"
+)
+
+// linearToS converts from a linear RGB component to an sRGB component.
+func linearToS(v float64) float64 {
+	if v > 0.0031308 {
+		return 1.055*math.Pow(v, 1/2.4) - 0.055
+	}
+	return 12.92 * v
+}
+
+// sToLinear converts from an sRGB component to a linear RGB component.
+func sToLinear(v float64) float64 {
+	if v > 0.04045 {
+		return math.Pow((v+0.055)/1.055, 2.4)
+	}
+	return v / 12.92
+}
+
+// linearRGB repesents a physically linear RGB color.
+type linearRGB struct {
+	R, G, B float64
+}
+
+// XYZ converts a linear RGB color to a CIE XYZ color.
+func (c linearRGB) XYZ() cieXYZ {
+	return cieXYZ{
+		X: 0.4124*c.R + 0.3576*c.G + 0.1805*c.B,
+		Y: 0.2126*c.R + 0.7152*c.G + 0.0722*c.B,
+		Z: 0.0193*c.R + 0.1192*c.G + 0.9505*c.B,
+	}
+}
+
+// linearRGB converts a CIEXYZ color to physically linear RGB.
+func (c cieXYZ) linearRGB() linearRGB {
+	return linearRGB{
+		R: c.X*3.2406 + c.Y*-1.5372 + c.Z*-0.4986,
+		G: c.X*-0.9689 + c.Y*1.8758 + c.Z*0.0415,
+		B: c.X*0.0557 + c.Y*-0.204 + c.Z*1.057,
+	}
+}
+
+// labTemp is an intermediate step in converting from CIE XYZ to CIE LAB.
+func labTemp(v float64) float64 {
+	if v > 0.008856 {
+		return math.Pow(v, 1.0/3.0)
+	}
+	return 7.787*v + 16.0/116.0
+}
+
+// LAB converts an CIE XYZ color to a CIE LAB color.
+func (c cieXYZ) LAB() cieLAB {
+	tempX := labTemp(c.X / 0.9505)
+	tempY := labTemp(c.Y)
+	tempZ := labTemp(c.Z / 1.089)
+	return cieLAB{
+		L: (116.0 * tempY) - 16.0,
+		A: 500.0 * (tempX - tempY),
+		B: 200 * (tempY - tempZ),
+	}
+}
+
+// sRGB represents a color within the sRGB color space, with an alpha channel
+// but not premultiplied.
+type sRGBA struct {
+	R, G, B, A float64
+}
+
+func (c sRGBA) linearRGB() linearRGB {
+	return linearRGB{
+		R: sToLinear(c.R),
+		G: sToLinear(c.G),
+		B: sToLinear(c.B),
+	}
+}
+
+// RGBA implements the color.Color interface.
+func (c sRGBA) RGBA() (r, g, b, a uint32) {
+	return uint32(c.R * c.A * 65535.0), uint32(c.G * c.A * 65535.0), uint32(c.B * c.A * 65535.0), uint32(c.A * 65535.0)
+}
+
+// LAB converts a sRGB color to a CIE LAB color.
+func (c sRGBA) LAB() cieLAB {
+	return c.linearRGB().XYZ().LAB()
+}
+
+// sRGB converts a CIE LAB color to an sRGBA color, where alpha is opacity
+// between 0 and 1.
+func (c cieLAB) sRGB(alpha float64) sRGBA {
+	return c.XYZ().linearRGB().S(alpha)
+}
+
+// colorTosRGBA converts a color to an sRGBA.
+func colorTosRGBA(c color.Color) sRGBA {
+	r, g, b, a := c.RGBA()
+	alpha := float64(a) / 65535.0
+	return sRGBA{
+		R: float64(r) / alpha / 65535.0,
+		G: float64(g) / alpha / 65535.0,
+		B: float64(b) / alpha / 65535.0,
+		A: alpha,
+	}
+}
+
+// ColorToMSH converts a color to MSH space.
+func ColorToMSH(c color.Color) MSH {
+	return colorTosRGBA(c).LAB().MSH()
+}
+
+// S converts a linear RGB color to an sRGBA color, where alpha is the alpha
```

Comment:
S is not exported because it's on an unexported type, and its role is to return an sRGBA, so rename to sRGBA.